bitcoin/src/blockdata/witness.rs

```diff
@@ -384,24 +384,72 @@ impl Witness {
-                    2
-                }
-            })
-            .filter(|&script_pos_from_last| len >= script_pos_from_last)
-            .and_then(|script_pos_from_last| self.nth(len - script_pos_from_last))
-            .map(Script::from_bytes)
+        self.last().and_then(|last| {
+            // From BIP341:
+            // If there are at least two witness elements, and the first byte of
+            // the last element is 0x50, this last element is called annex a
+            // and is removed from the witness stack.
+            if self.len() >= 3 && last.first() == Some(&TAPROOT_ANNEX_PREFIX) {
+                Some(Script::from_bytes(self.nth(self.len() - 3).unwrap()))
+            } else if self.len() >= 2 && last.first() != Some(&TAPROOT_ANNEX_PREFIX) {
+                Some(Script::from_bytes(self.nth(self.len() - 2).unwrap()))
+            } else {
+                None
+            }
+        })
+    }
+
+    /// Get the raw taproot control block following BIP341 rules.
+    ///
+    /// This does not guarantee that this represents a P2TR [`Witness`]. It
+    /// merely gets the last or second to last element depending on the first
+    /// byte of the last element being equal to 0x50. See
+    /// [Script::is_p2tr](crate::blockdata::script::Script::is_p2tr) to
+    /// check whether this is actually a Taproot witness.
+    pub fn raw_taproot_control_block(&self) -> Option<&[u8]> {
+        self.last().and_then(|last| {
+            // From BIP341:
+            // If there are at least two witness elements, and the first byte of
+            // the last element is 0x50, this last element is called annex a
+            // and is removed from the witness stack.
+            if self.len() >= 3 && last.first() == Some(&TAPROOT_ANNEX_PREFIX) {
+                Some(self.nth(self.len() - 2).unwrap())
+            } else if self.len() >= 2 && last.first() != Some(&TAPROOT_ANNEX_PREFIX) {
+                Some(last)
+            } else {
+                None
+            }
+        })
+    }
+
+    /// Get the taproot control block following BIP341 rules.
+    ///
+    /// This does not guarantee that this represents a P2TR [`Witness`]. It
+    /// merely gets the last or second to last element depending on the first
+    /// byte of the last element being equal to 0x50. See
+    /// [Script::is_p2tr](crate::blockdata::script::Script::is_p2tr) to
+    /// check whether this is actually a Taproot witness.
+    pub fn taproot_control_block(&self) -> Option<ControlBlock> {
+        self.raw_taproot_control_block().and_then(|cb| ControlBlock::decode(cb).ok())
+    }
+
+    /// Get the taproot annex following BIP341 rules.
+    ///
+    /// This does not guarantee that this represents a P2TR [`Witness`]. See
+    /// [Script::is_p2tr](crate::blockdata::script::Script::is_p2tr) to
+    /// check whether this is actually a Taproot witness.
+    pub fn taproot_annex(&self) -> Option<&[u8]> {
```

Comment:
We have an `Annex` type (in `sighash` for some reason), is there any reason why we don't use it here?

Reply:
That's kind of an empty type, right? It's clear that it's used to pass as an argument because it lives in sighash, and annexes have nothing to do with sighashes.

Reply:
The type basically just does a small invariant check that could be done inside the method that passes the annex. But yeah, I've always thought that the errors in the sighash module are actually quite annoying. I've been using it quite a lot recently and I think it merits some eyeballs.

Reply:
Looking at `Annex` it should've been unsized. But also I don't see a point in carrying around that extra byte. At best pushing it into `Witness` might be easier but really, I think there should be a trait for these things.

Anyway, I think semantically marking the bytes as `Annex` is a good idea not just for avoiding accidents but also allowing us to add methods to it once it gets used. So ACK changing the return type here.